### DISCUSSION TITLE
Server Node Event Source

### DISCUSSION THREAD

**Discussion by julian-gb** [FreeOpcUa/opcua-asyncio#1236] [General]
Hi, 

I tried to subscribe to events emitted by the server node. However, it seems events are not emitted by it but the [specification states](https://reference.opcfoundation.org/Core/Part3/v105/docs/4.6.1) "any OPC UA Server that supports eventing shall expose at least one Node as EventNotifier. The Server Object defined in OPC 10000-5 is used for this purpose." 

Is this not implemented or have I done something wrong? I was not able to find a disscussion regrading this topic. Has it maybe already dicussed somewhere?

**Comment by AndreasHeine** [comment]
it is!

you are using our client right!?
so there are some topics related to events in the past with the eventfilter!

but first things first does it work with ua expert?

**Comment by julian-gb** [comment]
Yes, tested it with the subscription examples and ua expert. Only when I subscribe directly to the emitting object I'm able to receive events.

Changing and running the `client-events.py` example while running the `server-events.py`.
```Python 
... 
        obj = await client.nodes.root.get_child(["0:Objects", "2:MyObject"])
        _logger.info("MyObject is: %r", obj)

        myevent = await client.nodes.root.get_child(["0:Types", "0:EventTypes", "0:BaseEventType", "2:MyFirstEvent"])
        _logger.info("MyFirstEventType is: %r", myevent)

        msclt = SubHandler()
        sub = await client.create_subscription(100, msclt)
        #handle = await sub.subscribe_events(obj, myevent)
        #handle = await sub.subscribe_events() # use default to subscribe to server base events
        handle = await sub.subscribe_events(client.nodes.server, myevent)
        await asyncio.sleep(10)
        await sub.unsubscribe(handle)
        await sub.delete()
...
```

Am I missing something?

**Reply by AndreasHeine** [reply]
ok i looked thru the source-code and the EventGenerator currently only support one emitting node so if you like to emit via server-obj then you need to modify the server-events.py

```python
# emmtiting node myobj
mysecondevgen = await server.get_event_generator(custom_etype, myobj)

# if no emmitting node is set it falls back to server
mysecondevgen = await server.get_event_generator(custom_etype)
```

```python
    async def get_event_generator(self, etype=None, emitting_node=ua.ObjectIds.Server):
        """
        Returns an event object using an event type from address space.
        Use this object to fire events
        """
        if not etype:
            etype = BaseEvent()
        ev_gen = EventGenerator(self.iserver.isession)
        await ev_gen.init(etype, emitting_node=emitting_node)
        return ev_gen
```

currently there is something missing:
The "Notifier-Chain" e.g. Server-Obj <-[HasNotifier-Reference or HasEventSource-Reference]- myobj
specreference: https://reference.opcfoundation.org/Core/Part3/v104/docs/7.18

so usually if implemented some day it should be possible to subscribe to server or myobj and recv in both cases the events of myobj

correct my if i am wrong! @oroulet @schroeder-

**Reply by schroeder-** [reply]
I think you are correct.

**Reply by julian-gb** [reply]
For now I tried to fix it by adding the following to the end of the `trigger` method in `event_generator.py`: 

```Python
     if self.emitting_node.nodeid is not ua.NodeId(ua.ObjectIds.Server):
          event = copy(self.event)
          event.emitting_node = ua.NodeId(ua.ObjectIds.Server)
          await self.isession.subscription_service.trigger_event(event)

     await self.isession.subscription_service.trigger_event(self.event)
````
Is the `HasNotifier` reference required in all cases? Seems to like a more sophisticated extension to the basic event functionality. Otherwise i would assume that the reference between server an emitting node should always been added if a new event generator is created, where the emitting node is not the server node.
